File: backend/app/features/extraction/schemas.py
from datetime import date

from pydantic import BaseModel, Field

from app.core.constants.limits import MAX_DESCRIPTION_LENGTH, MAX_OWNER_LENGTH
from app.core.enums import TaskPriority


class ExtractedTaskDTO(BaseModel):
    """Validated shape of a single task as returned by any AIProvider.

    This is the one place the LLM's output is trusted from — every field the
    model claims to have extracted is checked here (types, lengths, enum
    membership) before it can reach the database.
    """

    description: str = Field(min_length=1, max_length=MAX_DESCRIPTION_LENGTH)
    owner: str | None = Field(default=None, max_length=MAX_OWNER_LENGTH)
    due_date: date | None = None
    priority: TaskPriority = TaskPriority.MEDIUM
    source_text: str | None = None

File: frontend/src/features/tasks/schemas.ts
import { z } from "zod";
import { TaskPriority, TaskStatus } from "../../types/enums";

const PRIORITY_VALUES = Object.values(TaskPriority) as [TaskPriority, ...TaskPriority[]];
const STATUS_VALUES = Object.values(TaskStatus) as [TaskStatus, ...TaskStatus[]];

const MAX_DESCRIPTION_LENGTH = 2000;
const MAX_OWNER_LENGTH = 255;

export const taskEditSchema = z.object({
  description: z
    .string()
    .min(1, "Description is required.")
    .max(MAX_DESCRIPTION_LENGTH, "Description is too long."),
  owner: z.string().max(MAX_OWNER_LENGTH, "Owner name is too long.").optional(),
  priority: z.enum(PRIORITY_VALUES),
  status: z.enum(STATUS_VALUES),
  due_date: z.string().optional(),
});

export type TaskEditFormValues = z.infer<typeof taskEditSchema>;
